Consume CA render server sandbox extension earlier

pvollan opened the pull request:
#### 36cd8055de4c79d5b65002ecc250f74a6c16cc5a
<pre>
Consume CA render server sandbox extension earlier
<a href="https://bugs.webkit.org/show_bug.cgi?id=269757">https://bugs.webkit.org/show_bug.cgi?id=269757</a>
<a href="https://rdar.apple.com/123211653">rdar://123211653</a>

Reviewed by Brent Fulgham and Chris Dumez.

The sandbox extension needs to be consumed before the drawing area is created.

* Source/WebKit/WebProcess/WebPage/DrawingArea.cpp:
(WebKit::DrawingArea::create):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:

Canonical link: <a href="https://commits.webkit.org/275018@main">https://commits.webkit.org/275018@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1b46b07b84bf93d42f17013e85991d0c185d7e5b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40663 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19676 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43041 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43217 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36753 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22642 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17007 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33731 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41237 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16608 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35051 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14322 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14397 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44492 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36868 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36352 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40093 "Passed tests") | 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/40289 "Build is in progress. Recent messages:") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15478 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12691 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38441 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17097 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17148 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5407 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16741 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebKit/WebProcess/WebPage/WebPage.h:
```diff
@@ -2220,6 +2220,10 @@ class WebPage : public API::ObjectImpl<API::Object::Type::BundlePage>, public IP
     RefPtr<WebCore::Page> m_page;
 
+#if PLATFORM(MAC)
+    bool m_renderServerMachExtensionConsumed { false };
```
Where is this used?

Also, boolean variables should start with a prefix, maybe "was" in this case.

pvollan replied:
Ah, that is a good point. Actually, it was only declared to make sure that the sandbox extension is consumed before `m_drawingArea` was initialized. Alternatively, we could move the initialization of `m_drawingArea` into the body. What do you think?

Thanks for reviewing!

pvollan replied:
Another option is to consume the sandbox extension in the drawing area constructor.

cdumez replied:
Either option would be better than the proposal in the PR I think.

pvollan replied:
Yes, sounds good, I went for the last option :)

brentfulgham commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -514,6 +514,10 @@ static RefPtr<Frame> frameFromIdentifier(std::optional<FrameIdentifier> identifi
     : m_identifier(pageID)
+#if PLATFORM(MAC)
+    // This sandbox extension needs to be consumed before m_drawingArea is initialized.
+    , m_renderServerMachExtensionConsumed(SandboxExtension::consumePermanently(parameters.renderServerMachExtensionHandle))
```
I wonder what initialization requires this extension to exist. This feels a little fragile (what if someone tries to consolidate the bools together to save space!)

Maybe this should be an Optional or something to make it less attractive to merge with other bools, thereby breaking this ordering requirement.

brentfulgham commented on Source/WebKit/WebProcess/WebPage/DrawingArea.cpp:
```diff
@@ -54,6 +54,10 @@ using namespace WebCore;
 std::unique_ptr<DrawingArea> DrawingArea::create(WebPage& webPage, const WebPageCreationParameters& parameters)
 {
+#if PLATFORM(MAC)
+    SandboxExtension::consumePermanently(parameters.renderServerMachExtensionHandle);
```
Ah, this is excellent. Much safer! r=me

pvollan replied:
Thanks for reviewing!

pvollan committed:
Consume CA render server sandbox extension earlier
https://bugs.webkit.org/show_bug.cgi?id=269757
rdar://123211653

Reviewed by Brent Fulgham and Chris Dumez.

The sandbox extension needs to be consumed before the drawing area is created.

* Source/WebKit/WebProcess/WebPage/DrawingArea.cpp:
(WebKit::DrawingArea::create):
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:

Canonical link: https://commits.webkit.org/275018@main
